packages/react/src/utils.ts

```diff
@@ -0,0 +1,173 @@
+  (opts?: Partial<OperationContext>) => void
+];
+
+/**
+ * The inner result object returned from a mutation result
+ */
+export interface ActionHookState<Data = any, Variables = Record<string, any>> {
+  fetching: boolean;
+  stale: boolean;
+  data?: Data;
+  error?: ErrorWrapper;
+  extensions?: Record<string, any>;
+  operation?: Operation<Data, Variables>;
+}
+
+/**
+ * The return value of a `useAction`, `useGlobalAction`, `useBulkAction` etc hook.
+ * Includes the data result object and a function for running the mutation.
+ **/
+export declare type ActionHookResult<Data = any, Variables = Record<string, any>> = [
+  ActionHookState<Data, Variables>,
+  (variables?: Variables, context?: Partial<OperationContext>) => Promise<OperationResult<Data, Variables>>
+];
+
+const generateErrorMessage = (networkErr?: Error, graphQlErrs?: GraphQLError[]) => {
+  let error = "";
+  if (networkErr !== undefined) {
+    return (error = `[Network] ${networkErr.message}`);
+  }
+
+  if (graphQlErrs !== undefined) {
+    graphQlErrs.forEach((err) => {
+      error += `[GraphQL] ${err.message}\n`;
+    });
+  }
+
+  return error.trim();
+};
+
+const rehydrateGraphQlError = (error: any): GraphQLError => {
+  if (typeof error === "string") {
+    return new GraphQLError(error);
+  } else if (typeof error === "object" && error.message) {
+    return new GraphQLError(error.message, error.nodes, error.source, error.positions, error.path, error, error.extensions || {});
+  } else {
+    return error;
+  }
+};
+
+/**
+ * An error returned by any of the Gadget react hooks.
+ * Always has a message, but can be inspected to retrieve more detailed errors from either the network, the raw GraphQL layer, or Gadget specific errors like validation errors.
+ * Not intended for creating outside of Gadget-owned code.
+ **/
+export class ErrorWrapper extends Error {
+  /** @private */
+  static forClientSideError(error: Error, response?: any) {
+    return new ErrorWrapper({
+      executionErrors: [error],
+      response,
+    });
+  }
+  /** @private */
+  static forErrorsResponse(errors: Record<string, any>[], response?: any) {
+    return new ErrorWrapper({
+      executionErrors: errors.map(gadgetErrorFor),
+      response,
+    });
+  }
+  /** @private */
+  static forMaybeCombinedError(error?: CombinedError | null) {
+    if (!error) return undefined;
+    return new ErrorWrapper({
+      networkError: error.networkError,
+      executionErrors: error.graphQLErrors,
+      response: error.response,
+    });
+  }
+
+  /** @private */
+  static ensureData(result: UseQueryState<any>, dataPath: string[]) {
+    const nonNullableError = getNonNullableError(result, dataPath);
+    let error = ErrorWrapper.forMaybeCombinedError(result.error);
+    if (!error && nonNullableError) {
+      error = ErrorWrapper.forClientSideError(nonNullableError);
+    }
+    return error;
+  }
+
+  /** Class name of this error -- always `ErrorWrapper` */
+  public name: "ErrorWrapper";
```

Comment:
Two Qs:
1. Couldn't you just do `name = "ErrorWrapper"` here, instead of only assigning a value in the constructor?
2. What do you think of replacing this with a `get name() { return "ErrorWrapper" }` method instead? Just that should make it publicly consumable _and_ immutable, with very little more code.

Reply:
Can't do it the nice way cause TypeScript doesn't allow you to put statements before the `super` call in the constructor if you use this style. Is annoying.